cdap-examples/SpamClassifier/src/test/java/co/cask/cdap/examples/sparkstreaming/SpamClassifierTest.java

```diff
@@ -0,0 +1,186 @@
+import org.apache.twill.kafka.client.KafkaClientService;
+import org.apache.twill.kafka.client.KafkaPublisher;
+import org.apache.twill.zookeeper.ZKClientService;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Test for {@link SpamClassifier}
+ */
+public class SpamClassifierTest extends TestBase {
+
+  private static final String KAFKA_TOPIC = "someTopic";
+  private static final String KAFKA_BROKER_ID = "1";
+  private static ZKClientService zkClient;
+  private static KafkaClientService kafkaClient;
+
+  private static InMemoryZKServer zkServer;
+  private static EmbeddedKafkaServer kafkaServer;
+  private static int kafkaPort;
+
+  @BeforeClass
+  public static void init() throws Exception {
+    zkServer = InMemoryZKServer.builder().setDataDir(TMP_FOLDER.newFolder()).build();
+    zkServer.startAndWait();
+
+    kafkaPort = Networks.getRandomPort();
+    kafkaServer = new EmbeddedKafkaServer(generateKafkaConfig(zkServer.getConnectionStr(), kafkaPort,
+                                                              TMP_FOLDER.newFolder()));
+    kafkaServer.startAndWait();
+
+    zkClient = ZKClientService.Builder.of(zkServer.getConnectionStr()).build();
+    zkClient.startAndWait();
+
+    kafkaClient = new ZKKafkaClientService(zkClient);
+    kafkaClient.startAndWait();
+  }
+
+  @AfterClass
+  public static void cleanup() {
+    kafkaClient.stopAndWait();
+    zkClient.stopAndWait();
+    kafkaServer.stopAndWait();
+    zkServer.stopAndWait();
+  }
+
+  @Test
+  public void test() throws Exception {
+
+    // Deploy the KafkaIngestionApp application
+    ApplicationManager appManager = deployApplication(SpamClassifier.class);
+
+    ingestTrainingData();
+
+    publishKafkaMessages();
+
+    // start spark streaming program
+    SparkManager sparkManager = appManager.getSparkManager(SpamClassifierProgram.class.getSimpleName());
+    Map<String, String> runtimeArgs = new HashMap<>();
+    runtimeArgs.put("kafka.brokers", "127.0.0.1:" + kafkaPort);
+    runtimeArgs.put("kafka.topics", KAFKA_TOPIC);
+    sparkManager.start(runtimeArgs);
+
+    // Start and wait for service to start
+    final ServiceManager serviceManager = appManager.getServiceManager(SpamClassifier.SERVICE_HANDLER).start();
+    serviceManager.waitForStatus(true);
+
+    // wait for spark streaming program to write to dataset
+    Tasks.waitFor(HttpURLConnection.HTTP_OK, new Callable<Integer>() {
+      @Override
+      public Integer call() throws Exception {
+        return getResponse(serviceManager, "1").getResponseCode();
```

Comment:
I think you can just move the `testClassification` call in here. Rather then doing the assert inside the `testClassification` in the method, you can just have the method return `boolean` and have the `Tasks.waitFor` to expect a true.

```java
Tasks.waitFor(true, new Callable<Boolean>() {

...
return testClassification(...) && testClassification(...)
});